File: lib-segment/src/main/java/websiteschema/mpsegment/concept/Concept.java
package websiteschema.mpsegment.concept;

import java.util.ArrayList;
import java.util.List;

public class Concept {

    public static final Concept UNKNOWN = new Concept(0, "N/A");

    private long id;
    private String name;
    private List<Concept> children;
    private Concept parent;

    public Concept(long id, String name) {
        this.id = id;
        this.name = name;
        children = new ArrayList<Concept>();
    }

    public List<Concept> getChildren() {
        return children;
    }

    public Concept getParent() {
        return parent;
    }

    public List<Concept> getSiblings() {
        return getParent().getChildren();
    }

    public void addChild(Concept child) {
        child.setParent(this);
        children.add(child);
    }

    private void setParent(Concept parent) {
        this.parent = parent;
    }

    public String getName() {
        return name;
    }

    public long getId() {
        return id;
    }
}

File: lib-segment/src/main/scala/me/yingrui/segment/concept/Concept.scala
package me.yingrui.segment.concept

class Concept(id: Long, name: String) {

    private var description : String = null
    private var children = List[Concept]()
    private var parent: Concept = null

    def getChildren() : List[Concept] = {
        return children
    }

    def getParent() : Concept = {
        return parent
    }

    def getSiblings() : List[Concept] = {
        return getParent().getChildren()
    }

    def addChild(child: Concept) {
        child.setParent(this)
        children = child :: children
    }

    private def setParent(parent: Concept) {
        this.parent = parent
    }

    def getName() : String = {
        return name
    }

    def getId() : Long = {
        return id
    }

    def getDescription() : String = {
        return description
    }

    def setDescription(description: String) {
        this.description = description
    }
}

object Concept {
  val UNKNOWN = new Concept(0, "N/A")
}
